### ISSUE TITLE
optimize exclude_polygons computation

### ISSUE THREAD

**Issue by nilsnolde** [valhalla/valhalla#4387]
Leaving it here for future work:

Instead of collecting all edges within each polygon and adding them to the user-avoided edges, we could collect only the edges _crossing_ a polygon ring. On a request, we'd then first check if the origin/destination correlated points are inside those polygons or not, write that info to the EdgeLabel as a boolean and switch the boolean once we encounter one of those crossed edges. That would have many advantages:
- make things a bit faster for small polygons and a LOT (exponentially) faster for big polygons
- EdgeLabel structure is easy to extend
- we could even track more than one property easily, not only to exclude but also other properties
- having it on the labels would carry the info to Odin, which could be very useful once we decided to modify costing for edges inside polygons (instead of only excluding)
- it'd be trivial to enable inner rings

**Comment by nilsnolde** [comment]
Also the lowest hanging fruit of all: I recently discovered that boost::geometry doesn't do any bbox check in its predicates. Doing that very basic thing should already speed this up many times.

**Comment by ADMUKHE** [comment]
Does the multiple list of polygons in the exclude_polygon argument mean the area of intersection between the several rings? or the rings can be seperate polygons in itself?

**Comment by ADMUKHE** [comment]
If i understood it correctly, The "max_exclude_polygon_length" seems configurable in the code, so just changing this value does not require any change in the code?  However if we want the performance to be as good or better or introduce inner rings then we need to compute this in the approach you mentioned above right?

**Comment by jcoupey** [comment]
> boost::geometry doesn't do any bbox check in its predicates

That would be a good way to not pay the price for all PIP checks in most cases. But if those checks are a bottleneck you may also consider storing your predefined polygons in a [R-tree structure](https://en.wikipedia.org/wiki/R-tree). That would further reduce the number of checks (including simple bbox checks) by an order of magnitude. There is even a R-tree structure in `boost::geometry` that I remember using a while back for exactly that purpose.

**Comment by nilsnolde** [comment]
Hehe yeah that’s not an option unfortunately. We’d have to put ALL edges in an index, that takes ages and consumes crazy amounts of memory. Could be optional, but wouldn’t like to see that code😄We’re too flexible for our own good sometimes;)

**Comment by nilsnolde** [comment]
@ADMUKHE right. The perf would be one aspect, the more interesting thing for me would be that it opens up a ton of more options  what to actually do with polygons. We could get really creative at that point.

**Comment by jcoupey** [comment]
> We’d have to put ALL edges in an index

Only the polygons (or some kind of reference to them) would be stored in a R-tree based on their bbox. I don't really get how this would require that much memory, but then I may be misunderstanding the use-case or the other constraints you have! ;-)

**Comment by nilsnolde** [comment]
There's typically only a few polygons, but millions of edges, so that wouldn't buy us much. In fact, our tiles act as some kind of poor man's spatial index with a size of 0.05°. It's better than nothing but also really not good enough for those kind of workloads.
